scipy/cluster/hierarchy.py

```diff
@@ -1391,6 +1391,170 @@ def cut_tree(Z, n_clusters=None, height=None):
     return groups.T
 
 
+def cut_tree_balanced(Z, max_cluster_size):
+    """
+    Given a linkage matrix Z, return the balanced cut tree (where each
+    resulting cluster has a maximum size of max_cluster_size).
+
+    The function looks recursively along the hierarchical tree, from the root
+    (single cluster gathering all the samples) to the leaves (i.e. the clusters
+    with only one sample), retrieving the biggest possible clusters containing
+    a number of samples lower than a given maximum. If a cluster at a specific
+    tree level contains a number of samples higher than the given maximum, it
+    is ignored and its offspring (smaller) sub-clusters are taken into
+    consideration. If the cluster contains a number of samples lower than the
+    given maximum, it is taken as result and its offspring sub-clusters not
+    further processed.
+
+    Parameters
+    ----------
+    Z : scipy.cluster.linkage array
+        The linkage matrix resulting (for instance) from calling the method
+        scipy.cluster.hierarchy.ward(). I.e. it contains the hierarchical
+        clustering encoded as a linkage matrix.
+
+    max_cluster_size : array_like
+        maximum number of data samples contained within the resulting clusters.
+        Thus, all resulting clusters will contain a number of data samples
+        <= max_cluster_size. Note that max_cluster_size must be >= 1.
+
+    Returns
+    -------
+    cluster_id : ndarray
+        one-dimensional numpy array of integers containing for each input
+        sample its corresponding cluster id. The cluster id is an integer which
+        is higher for deeper tree levels.
+
+    cluster_level : ndarray
+        one-dimensional numpy array of strings containing for each input sample
+        its corresponding cluster tree level, i.e. a string of '0's and '1's
+        separated by '.' Note that the cluster level is longer for deeper tree
+        levels, being 0 the root cluster, 0.0 and 0.1 its offspring, and so on.
+
+    Examples
+    --------
+    >>> from scipy import cluster
+    >>> from scipy import stats
+
+    Initialize the random seed.
+
+    >>> np.random.seed(14)
+
+    Create a input matrix containing 100 data samples with 4 dimensions. Note:
+    using gamma() in order to generate an unbalanced distribution. If a regular
+    cut_tree() would be performed, one big and many small clusters would be
+    obtained.
+
+    >>> X = stats.gamma.rvs(0.1, size=400).reshape((100,4))
+
+    Compute the linkage matrix using the scipy ward() method.
+
+    >>> Z = cluster.hierarchy.ward(X)
+
+    Perform a balanced cut tree of the linkage matrix.
+
+    >>> [cluster_id, cluster_level] = cut_tree_balanced(Z, max_cluster_size=10)
+    >>> cluster_id[:10]
+    >>> cluster_level[:10]
```

Comment:
I am not very sure about this suggestion. Although at first glance it is a sensible suggestion, I think the example should work also in a script, and not only in a jupyter notebook. Further, the dendrogram plot does not show the cluster cut, which is the main relevant issue within this present function. Therefore, in my humble opinion, the inclusion of the dendrogram() function within the example would be rather misleading for the readers / users. Anyhow, if this inclusion is strongly suggested, I will proceed with this amendment in the next version of the PR.

Reply:
> I think the example should work also in a script, and not only in a jupyter notebook.

Matplotlib plots work just fine in scripts. We have many plotting examples in the docs.

> the dendrogram plot does not show the cluster cut, which is the main relevant issue within this present function.

That's a good reason indeed. I think it's still possible to visualize something here, for example http://www.sthda.com/english/wiki/beautiful-dendrogram-visualizations-in-r-5-must-known-methods-unsupervised-machine-learning uses the cuts to color a dendrogram. Let's leave that as a bonus exercise though.

Reply:
Actually, now that my memory is refreshed: the README at https://github.com/vreyespue/cut_tree_balanced does contain nice graphical examples.

Reply:
It would have been nice to add those graphical examples to the description of this PR, but as I wrote this is not possible (I just added some links to the original repo). 
In this regard, now I am including a link to the original repo in the Notes section of the docstring, so that the user can refer to those graphical examples easily.
As a bonus, please not that the jupyter notebook used to create those graphical examples [is now available](https://github.com/vreyespue/cut_tree_balanced/tree/master/paper/figures_dendrograms). There you can see that the process to create those examples was manual, and adapted to the specific example.
Many thanks and best regards.